### PR TITLE
basic big int implementation

### PR DESCRIPTION
**Pull request by b05702057**

### CODE REVIEW

File: design.md [hunk 1]
```diff
@@ -1,122 +1,210 @@
...
+10.
+List Test (for when it’s implemented): 
+“x:int = 900000001
+list1 = [3] * x
+print(list1[x-1])”
+Expected: 3
+
+The below is related to the concerns from instructors.
+
+(1)
```
Comment (GV1028) [inline design.md]: Thanks for explaining all the changes in each file!

File: tests/import-object.test.ts [hunk 2]
```diff
@@ -13,9 +14,115 @@ function stringify(typ: Type, arg: any): string {
...
+      break
+  }
+
+  var isNegative = 1;
+  if (bigInt3 < 0) {
+    isNegative = -1;
+    bigInt3 *= BigInt(-1);
+  }
+
+  var i = 1; // the first field is preserved for the size
```
Comment (GV1028) [inline tests/import-object.test.ts]: You might want to make this in to another helper function (createBigInt)

File: design.md [hunk 3]
```diff
@@ -1,122 +1,210 @@
...
+“x:int = 5000000000000000000000000000 - 2000000000000000000000000000
+print(x)” 
+Expected: 3000000000000000000000000000
+6.
+Multiply i32s to get BigNum: 
+“x:int = 10000*300000
+print(x)” 
+Expected: 3000000000
+7.
+Divide:
```
Comment (GV1028) [inline design.md]: Add test: Divide bignums to get i32

File: design.md [hunk 4]
```diff
@@ -1,122 +1,210 @@
...
+print(x)” 
+Expected: “999999999999999999999999999999999999999999999999” 
+3.
+Unary operation: 
+“x: int = -999999999999999999999999999999999999999999999999 
+print(x)” 
+Expected: “-999999999999999999999999999999999999999999999999”
+4.
+Basic addition: 
+“x:int = 1000000000000000000000000000 + 1000000000000000000000000000
```
Comment (GV1028) [inline design.md]: Would like to see tests on bignums where the majority of the digits are not 0's.

File: design.md [hunk 5]
```diff
@@ -1,122 +1,210 @@
...
+Test for initial compilation: 
+“999999999999999999999999999999999999999999999999” 
+Expected: 999999999999999999999999999999999999999999999999
+2.
+Basic assignment and printing: 
+“x: int = 999999999999999999999999999999999999999999999999 
+print(x)” 
+Expected: “999999999999999999999999999999999999999999999999” 
+3.
+Unary operation:
```
Comment (GV1028) [inline design.md]: Add tests for negative big ints turning to positive big ints

**Review by GV1028 — COMMENTED** [review]
Overall, this is a good design! Great job! I really like the idea of representing bignums as a class. 
The tests also look good, you might want to add just a few more tests for unaryop and bignums with other digits than 0.
It's worth talking to the memory management group and the lists/strings groups about how numbers will be represented and work as offsets when they are bignums. It might be that you collaborate a little bit to provide a helper in that case.

File: design.md (hunk 1 above)
Comment (jmakings) [reply design.md]: Thanks Vignesh for your constructive comments, we'll try to implement all the changes you suggested in the coming days. We'll let you know if we have any questions/clarifications.

File: design.md (hunk 5 above)
Comment (b05702057) [reply design.md]: Hi @GV1028 , 
  Thanks for your suggestion! I've updated all implementations and tests except for the part that is related to other teams. Also, I've been discussing with the string, list, and memory management team. Do you think it is possible that our branch get merged first, so they can start modify their part based on the new environment?

File: lower.ts [hunk 6]
```diff
@@ -267,7 +267,7 @@ function flattenExprToExpr(e : AST.Expr<[Type, SourceLocation]>, env : GlobalEnv
       const classdata = env.classes.get(e.name);
       const fields = [...classdata.entries()];
       const newName = generateName("newObj");
-      const alloc : IR.Expr<[Type, SourceLocation]> = { tag: "alloc", amount: { tag: "wasmint", value: fields.length } };
+      const alloc : IR.Expr<[Type, SourceLocation]> = { tag: "alloc", amount: { tag: "wasmint", value: BigInt(fields.length) } };
```
Comment (jpolitz) [inline lower.ts]: It's surprising to me that this change is made to the argument to alloc, but there's no change to the _implementation_ of alloc. What happens if alloc gets an actual big number? Aren't bignums heap allocated, won't that affect how alloc works?

File: ir.ts [hunk 7]
```diff
@@ -34,7 +34,7 @@ export type Expr<A> =
 
 export type Value<A> = 
     { a?: A, tag: "num", value: bigint }
-  | { a?: A, tag: "wasmint", value: number }
+  | { a?: A, tag: "wasmint", value: bigint }
```
Comment (jpolitz) [inline ir.ts]: I'm confused about this. We need _something_ that represents plain numbers in generated code, for example small offsets to fields. If we actually put in a bigint here, it would produce non-well-formed WASM code with a literal that is too large.

File: lower.ts [hunk 8]
```diff
@@ -3,26 +3,26 @@ import * as IR from './ir';
 import { Type, SourceLocation } from './ast';
 import { GlobalEnv } from './compiler';
 
-const nameCounters : Map<string, number> = new Map();
+const nameCounters : Map<string, bigint> = new Map();
```
Comment (jpolitz) [inline lower.ts:7]: Again, this change suggests that we have programs with more than `2^32` (or even more) variables!

Comment (b05702057) [reply lower.ts:7]: Hi Joe,

  Thanks for the questions. I have fixed the above code. Initially I make all numbers javaScript big numbers because I don't want to be confused by when to use numbers and when to use big numbers, but big numbers are indeed overkill in these cases and could cause confusion for others who don't understand our design.

Sincerely,
Zack
